### PR TITLE
Refactor web::latest_version, move logic into CrateDetails

### PR DESCRIPTION
**Pull request by koenaad**
I noticed `web/mod.rs` contained a really complicated `latest_version` function: https://github.com/rust-lang/docs.rs/blob/master/src/web/mod.rs#L316
The bulk of the work was spent in converting the version strings into `semver::Version`, to sort the array. Something that is already done in `CrateDetails`...

The original `latest_version` would return:
- `None` if `req_version` was an invalid function or if `req_version` is the latest version
- `Some(version)` if this crate has a more recent version

I have split what this function does in two functions:
- `CrateDetails::is_latest_version` --> check if this is the most recent release
- `CrateDetails::latest_version` --> grab the most recent release

I like how it is much easier to reason about now 😊

These changes should be easy to follow commit-by-commit. The real changes happen in the second commit: 309ea7a

### CODE REVIEW

File: src/web/rustdoc.rs [hunk 1]
```diff
@@ -262,7 +262,9 @@ pub fn rustdoc_html_server_handler(req: &mut Request) -> IronResult<Response> {
 
     content.full = file_content;
     let crate_details = cexpect!(CrateDetails::new(&conn, &name, &version));
-    let (path, latest_version) = if let Some(latest_version) = latest_version(&crate_details.versions(), &version) {
+
+    let (path, latest_version) = if !crate_details.is_latest_version() {
+        let latest_version = crate_details.latest_version().to_string();
```
Comment (jyn514) [inline src/web/rustdoc.rs]: This now calculates the latest version twice. Not a giant deal since it's now just a field, but it would be nice to only calculate once.

File: src/web/crate_details.rs [hunk 2]
```diff
@@ -255,14 +255,17 @@ impl CrateDetails {
...
+    /// Returns whether this release is the latest release of this crate.
+    pub fn is_latest_version(&self) -> bool {
+        self.version == self.latest_version()
     }
-}
 
+    /// Returns the version of the latest release of this crate.
+    pub fn latest_version(&self) -> &str {
+        // this is safe to unwrap: releases will always contain at least the version of this release
+        &self.releases.get(0).unwrap().version
```
Comment (jyn514) [inline src/web/crate_details.rs]:
```suggestion
        &self.releases[0].version
```

**Comment by koenaad** [comment]
@jyn514 I have simplified the code a bit, avoiding the need to call `latest_version()` twice. Can you take a look again? :)

File: src/web/mod.rs [hunk 3]
```diff
@@ -547,16 +513,4 @@ mod test {
...
-    #[test]
-    fn test_latest_version() {
-        let versions = vec!["1.0.0".to_string(),
-                            "1.1.0".to_string(),
-                            "0.9.0".to_string(),
-                            "0.9.1".to_string()];
-        assert_eq!(latest_version(&versions, "1.1.0"), None);
-        assert_eq!(latest_version(&versions, "1.0.0"), Some("1.1.0".to_owned()));
-        assert_eq!(latest_version(&versions, "0.9.0"), Some("1.1.0".to_owned()));
-        assert_eq!(latest_version(&versions, "invalidversion"), None);
```
Comment (jyn514) [inline src/web/mod.rs]: Please add these test cases to the new `test_latest_version` function. It would also be great if you could add a test case that these are sorted by semver instead of alphabetically: `1.12.0` comes lexicographically before `1.2.0`, but should come after when sorted by version.

**Comment by jyn514** [comment]
Overall I like this :) It will make https://github.com/rust-lang/docs.rs/issues/396, https://github.com/rust-lang/docs.rs/issues/312, https://github.com/rust-lang/docs.rs/issues/223, and https://github.com/rust-lang/docs.rs/issues/395 easier to fix now that everything's in one place.

#395 would be a good test case for this actually, let me run that locally.

**Comment by jyn514** [comment]
395 was a separate issue in `match_version`, it won't be affected.

File: src/web/mod.rs (hunk 3 above)
Comment (koenaad) [reply src/web/mod.rs]: Good idea! I have added these edge-cases to `test_releases_should_be_sorted` since this test focuses solely on `releases` being sorted correctly.
I would prefer not to put the same test cases in `test_latest_version` to avoid having duplicate tests.

Is that acceptable for you?

**Comment by koenaad** [comment]
> Overall I like this :) It will make #396, #312, #223, and #395 easier to fix now that everything's in one place.
> 
> #395 would be a good test case for this actually, let me run that locally.

About those issues mentioning yanked crates, I have another small-ish refactor coming up which would add a `yanked` bool to `Release`. This should make it fairly easy to improve `latest_version()` and ignore yanked crates.

File: src/web/mod.rs (hunk 3 above)
Comment (jyn514) [reply src/web/mod.rs]: Looks good to me :)
